core/typeSugar.ml

```diff
@@ -2950,10 +2951,21 @@ let rec type_check : context -> phrase -> phrase * Types.datatype * usagemap =
             let () = unify ~handle:Gripers.query_outer
               (no_pos (`Record context.effect_row), no_pos (`Record outer_effects)) in
             let p = type_check (bind_effects context inner_effects) p in
-            let () = if Settings.get  Database.shredding then ()
-                     else let shape = Types.make_list_type (`Record (StringMap.empty, Types.fresh_row_variable (lin_any, res_base), false)) in
-                          unify ~handle:Gripers.query_base_row (pos_and_typ p, no_pos shape) in
-            Query (range, erase p, Some (typ p)), typ p, merge_usages [range_usages; usages p]
+            let evaluator =
+              match policy with
+                | Nested -> `Nested
+                | Plain -> `Plain
+                | Default -> if (Settings.get Database.shredding) then `Nested else `Plain in
+            let () =
+              match evaluator with
+                | `Nested -> ()
+                | `Plain  ->
+                     let shape =
+                       Types.make_list_type
+                         (`Record (StringMap.empty,
+                            Types.fresh_row_variable (lin_any, res_base), false)) in
+                     unify ~handle:Gripers.query_base_row (pos_and_typ p, no_pos shape) in
+            Query (range, policy, erase p, Some (typ p)), typ p, merge_usages [range_usages; usages p]
```

Comment:
There are a couple of dark corners here that could potentially be improved:

* Plain queries allow `orderby` and correctly handle converting regex matching to `LIKE`, whereas shredding doesn't try to handle the former and fails to handle the latter

* Plain queries also allow ranges.  For nested queries, these are just ignored.

* The query keyword does double duty as "here is a query; run it" and "this code is expected to be tame".  It's unclear to me what happens, or what should happen, if `query plain {...}` is nested inside `query nested {...}` or vice versa.  If this should never be possible then I don't see how to stop this using the current type system.

I'm not sure it makes sense to try to handle these all statically, but it would be good to at least check these corner cases and see if the resulting error messages are comprehensible, and if not improve them.

Reply:
Thanks, @jamescheney: 

> Plain queries allow orderby and correctly handle converting regex matching to LIKE, whereas shredding doesn't try to handle the former and fails to handle the latter

> Plain queries also allow ranges. For nested queries, these are just ignored.

I think that given this additional static information, it's perfectly possible to make these forms either a parse or a type error. I will do this.

> The query keyword does double duty as "here is a query; run it" and "this code is expected to be tame". It's unclear to me what happens, or what should happen, if query plain {...} is nested inside query nested {...} or vice versa. If this should never be possible then I don't see how to stop this using the current type system.

Yes, I was wondering about this. I had some thoughts about tracking it in the effect system, but it felt fiddly. I think the best way forward for this patch is to pass the top-level policy into the normaliser and error out if a block with a different policy is detected -- this will rule out possible strange behaviours or bad error messages.

Reply:
At the moment, I think everything allowed in a nested query is also allowed in a flat/ordered query, so adding an effect that flags use of orderby might work, but would entail many changes.

Longer term, I have some hope that we can unify the two approaches so that we don't need two policies. But I've been saying that for years now, and having a workable approach that supports different query backends would be a good idea anyway. (For example to handle queries for different data models).

Another strategy could be to distinguish between the "run" behavior of the query keyword and the "quote" behavior, which might be useful for other things. Then either dynamically error if we try to run a query with unhandled features (like in F#) or index the quoted code type to keep track of what language subset is used.  This feels a bit related to algebraic effects and handlers.